[feature] 지도 화면 영역에 존재하는 픽 정보 마커 표시

meanjoo opened the pull request:
## #️⃣연관된 이슈
- resolved: #30 

## 📝작업 내용 및 코드
- 지도 화면 영역에 맞게 픽 불러오기
- MapScreen으로 전환 시 마지막 카메라 위치 유지

https://github.com/user-attachments/assets/a4b3d05c-182e-41c8-8781-73b5451f6e49

### 지도 화면 영역에 맞게 픽 불러오기
카메라가 이동하면 카메라 변경 이벤트가 발생하고, 카메라 움직임이 끝나 대기 상태가 되면 카메라 대기 이벤트가 발생한다.
카메라 대기 이벤트가 발생했을 때 화면에 보이는 영역을 계산한 후 이 영역에 포함된 픽을 불러오면 된다고 생각했다. 카메라 변경 이벤트에서 영역을 계산하고 픽을 불러온다면 거의 실시간으로 픽을 업데이트 할 수 있겠지만 너무 잦은 요청을 할 것 같다.

화면에 보이는 지도의 좌표를 구하기 위해 화면의 픽셀을 구해서 진행해야 하나 했는데 [NaverMap 객체](https://navermaps.github.io/android-map-sdk/reference/com/naver/maps/map/NaverMap.html) 안에 있는 값 중에 contentWidth와 contentHeight가 있었다. contentWidth는 패딩을 제외한 지도 뷰의 화면 상 너비(픽셀 단위), contentHeight는 패딩을 제외한 지도 뷰의 화면 상 높이(픽셀 단위)이다. 이 값을 이용하여 화면의 왼쪽 상단, 오른쪽 하단의 지도 좌표를 구했다.
```kotlin
val leftTop = projection.fromScreenLocation(PointF(0f, 0f))
val rightBottom = projection.fromScreenLocation(PointF(contentWidth.toFloat(), contentHeight.toFloat()))
```

딱 화면의 네모 영역만큼은 어떻게 불러와야할지 모르겠어서 기존에 있던 중심 좌표와 반경을 이용해서 중심 좌표로부터 반경 안에 존재하는 픽을 불러오는 함수를 활용하기로 했다. 화면의 왼쪽 상단 지도 좌표와 오른쪽 하단 지도 좌표를 알고 있으면 이 둘을 이용해서 중심 좌표를 구할 수 있고, 거리를 /2 한 값으로 반경도 구할 수 있다. 이렇게 하면 빨간색 영역만큼의 픽을 불러올 수 있다.

<img src="https://github.com/user-attachments/assets/be8c8be9-87cf-4a60-a825-dde20da99563" width="300">

기존에 존재하는 fetchPickInArea()를 활용할 수 있는지 체크했다. 기존에 작성된 함수 안을 보면 bounds를 구하는 곳에서 `GeoFireUtils.getGeoHashQueryBounds()`를 사용하고 있다.

```kotlin
override suspend fun fetchPicksInArea(
        lat: Double,
        lng: Double,
        radiusInM: Double
    ): List<Pick> {
        val center = GeoLocation(lat, lng)
        val bounds = GeoFireUtils.getGeoHashQueryBounds(center, radiusInM)
        /* ... */
```
<img src="https://github.com/user-attachments/assets/42db2f79-7a53-404e-86cd-32f32f77c50b" width="600">

이는 데이터베이스의 쿼리의 경계로 사용할 시작과 종료 geohash를 결정하는 함수인데, 인자로 받는 radius를 최대 8587km까지 지원한다고 하기 때문에 이를 만족하는지 확인해야 했다.

화면에 보이는 지도 영역의 반경은 GeoFireUtils에서 제공하는 `getDistanceBetween()`을 2로 나눈 값으로 구한다.

<img src="https://github.com/user-attachments/assets/79b6ef0a-6747-4094-ae88-260f83de015f" width="500">

지도를 줌을 최소로 하여 움직였을 때 `getDistanceBetween()`으로 얻어지는 두 좌표 사이의 미터 단위 거리 최대 값이 1,xxx,xxx(m)이고 이를 km 단위로 바꾸면 1,xxx(km)이다.

![image (8)](https://github.com/user-attachments/assets/2068700f-afaa-423a-a969-c6ab68773f6e)

두 좌표 사이의 거리가 8587km보다 작기 때문에 이 값을 2로 나눈 반경도 당연히 8587km보다 작기 때문에 `GeoFireUtils.getGeoHashQueryBounds()`를 사용할 수 있다고 판단했다. 따라서 지도 화면 영역에 존재하는 픽을 불러오는 것은 화면의 왼쪽 상단, 오른쪽 하단의 지도 좌표를 구한 후 이 값을 가지고 기존의 함수를 활용하는 방향으로 구현했다.

### MapScreen으로 전환 시 마지막 카메라 위치 유지
지도를 움직여서 픽을 선택한 후 상세 정보를 확인한 후 다시 지도로 돌아오면 카메라의 위치는 현위치로, 줌 레벨은 프로젝트에서 설정한 기본 줌 레벨로 고정되어 있었다. 다시 돌아와도 마지막 본 화면을 유지하기 위해 아래와 같이 작업을 했다.

뷰모델에 마지막 카메라 위치를 가지고 있는 변수 lastCameraPosition을 만들고 NaverMap의 onDispose에서 현재 카메라 위치를 뷰모델의 lastCameraPosition에 저장하도록 했다. 기존에는 기기 위치를 설정하는 함수에서 현재 위치로 카메라를 이동시키고 줌 레벨을 설정했었는데, 이 함수에서 만약 lastCameraPosition이 null이 아니라면 카메라를 이전 카메라 위치로 이동시키도록 했다.


## 💬리뷰 요구사항
- 에뮬레이터나 다른 기기에서도 잘 동작하는지 테스트해주시면 감사하겠습니다.
  - 화면 영역에 있는 픽들이 잘 불러와지는지
  - 검색이나 상세 정보를 보고 지도로 돌아온 후에도 이전 카메라 위치가 유지되는지

meanjoo committed:
[feature] 화면 영역에 맞게 픽 불러오기

meanjoo committed:
[refactor] 사용하지 않는 함수 삭제

meanjoo committed:
[feature] MapScreen으로 전환 시 마지막 카메라 위치 유지

meanjoo committed:
[refactor] 사용하지 않는 변수 제거

meanjoo committed:
[refactor] 클릭의 enabled 속성으로 사용자 경험 개선

상호작용이 불가능한 상태에서도 클릭되는 효과가 있던 것을 제거

yuni-ju reviewed (approved):
에뮬레이터로 위치도 바꿔보고 이것저것 동작해봐도 반경에 따라 픽 불러오기와 마지막 카메라 위치 유지 잘 되네요 ㅎㅎ

처음에는 leftTop, rightBottom용으로만 쓰이는 lat1, lat2라는 파라미터가 무엇을 의미하는 지 헷갈릴 수도 있겠다고 생각했으나 추후 재사용성을 고려하면 괜찮을 것 같다는 생각을 했습니다! 

반경 설정하신 아이디어도 좋고 클릭 enable하신 것도 확인했습니다! 고생하셨습니다 👍 👍

miller198 commented on app/src/main/java/com/squirtles/musicroad/map/BottomNavigation.kt:
```diff
@@ -88,8 +88,8 @@ fun BottomNavigation(
                 )
-                .clickable {
-                    if (lastLocation != null) onCenterClick()
+                .clickable(enabled = lastLocation != null) {
```
이거 몰랐네요 ㅋㅋㅋㅋㅋ 이렇게하면되는군요

miller198 commented on domain/src/main/java/com/squirtles/domain/usecase/FetchPicksInBoundsUseCase.kt:
```diff
@@ -0,0 +1,11 @@
+import com.squirtles.domain.repository.FirebaseRepository
+import javax.inject.Inject
+
+class FetchPicksInBoundsUseCase @Inject constructor(
```
기존에 MapViewModel에 있던 거리계산 함수를 사용해서 화면 범위 기준 radius 를 MapViewModel에서 계산하고
유스케이스와 레포지토리,데이터소스에 함수 추가 없이 기존 FetchPickInAreaUseCase를 재활용 할 수 있었을 것 같네요

meanjoo replied:
그렇네요 굳이 따로 UseCase를 만들 필요가 없이 뷰모델에서 이를 처리하면 기존의 UseCase를 활용할 수 있네요

기존의 UseCase를 활용하게 되면 거리 계산을 뷰모델에서 해야하는데 그럼 거리 계산을 위해 사용하는 함수가 달라지게 됩니다. UseCase를 따로 만들었을 때는 GeoFireUtils에서 제공하는 함수를, 뷰모델에서 하면 LatLng에서 제공하는 함수를 사용해야 합니다. 당연히 둘 다 거리 계산을 하는 거라 비슷한 값이 나오겠지만 혹시 몰라 로그를 찍어봤는데 무시할 수 있을만한 오차라 생각하여 기존의 UseCase를 활용하는 방식으로 변경했습니다.

![image (5)](https://github.com/user-attachments/assets/32174fb2-598f-447e-921c-16dc8db12a48)

기존에 있는 거리 계산 함수는 LatLng을 사용하지 않아서 이를 활용하지는 않고 `val radiusInM = leftTop.distanceTo(rightBottom)` 이렇게 LatLng 안에 있는 distanceTo를 이용해서 거리를 쟀습니다. 기존의 거리 계산 함수는 Location을 사용해서 LatLng에서 제공하는 함수를 사용했습니다.

miller198 replied:
좋습니다~고생하셨어요

meanjoo committed:
[refactor] 기존의 UseCase 활용

지도 화면 영역의 픽을 불러오는 데에 새로운 UseCase를 만들어서 사용하지
않고 특정 영역에 존재하는 픽을 불러오는 기존의 UseCase를 활용
